docs/design/operator.md

```diff
@@ -0,0 +1,407 @@
+management interfaces for Ceph CSI drivers. It provides a set of new Kubernetes
+APIs (CRDs) for governing deployment and management of Ceph-CSI drivers, as
+well as connection details to Ceph clusters. The operator ensures that the
+desired state specified in the CRDs is maintained in the cluster.
+
+## Assumptions and Dependencies
+
+Kubernetes cluster is available and operational. Ceph cluster is set up and
+accessible from the Kubernetes cluster. Users are familiar with Kubernetes and
+Ceph concepts. CRD and operator versions are compatible with the Kubernetes
+version in use.
+
+## General Constraints
+
+Resource limitations of the Kubernetes cluster (CPU, memory). Network latency
+and connectivity between the Kubernetes cluster and Ceph cluster. Watching for
+changes on the ceph cluster and updating internal details.  Security policies
+and RBAC configurations in the Kubernetes cluster.
+
+## Goals and Guidelines
+
+- Automate the deployment and lifecycle for Ceph-CSI drivers.
+- Provide a consistent and flexible configuration mechanism using CRDs.
+- Enable namespace-scoped configurations for different Ceph-CSI drivers.
+- Ensure high availability and scalability of the operator.
+- Support multiple deployment methods for the operator (single-namespace,
+  multi-namespace, all-namespaces).
+- Support multiple operator deployment in a single cluster when running in a
+  single-namespace mode.
+
+## Development Methods
+
+Use operator framework which is higly used in operator developed and also by
+following Kubernetes best practices for operator development.
+
+## Architecture
+
+```mermaid
+graph TD
+    A[CRD Changes] --> B[Operator]
+    B -->  D[Configure CephFS CSI Driver]
+    B -->  E[Configure RBD CSI Driver]
+    B -->  F[Configure NFS CSI Driver]
+
+```
+
+In this updated diagram:
+
+- **CRD Changes**: Represents changes made to Custom Resource Definitions,
+  which trigger actions in the operator.
+- **Operator**: Listens for changes in CRDs and initiates the installation and
+  configuration of CSI drivers.
+- **Configure CephFS, NFS, RBD**: Actions performed by the operator to install
+  and configure the respective CSI drivers based on the CRD changes.
+
+## CRD's for ceph-csi-operator
+
+### CephCSIOperatorConfig CRD
+
+Manages operator-level configurations and offers a place to overwrite settings
+for CSI drivers. This CRD is a namespace-scoped CRD and a single CR named
+ceph-csi-operator-config can be created by the admin in the namespace
+where operator is running. The configurations are catagorized into 2 different
+parts
+
+- Operator configurations
+- Common configurations across all CSI driver
+
+```yaml
+---
+kind: CephCSIOperatorConfig
+apiVersion: csi.ceph.io/v1alpha1
+metadata:
+  name: csioperatorconfig
+  namespace: operator-namespace
+spec:
+  logLevel: 1
+  driverSpecDefaults:
+    logging:
+      logLevel: 5
+      maxfiles: 5
+      maxLogSize: 10M
+    clusterName: 5c63ad7e-74fe-4724-a511-4ccdc560da56
```

Comment:
Is this the ceph fsid?
```suggestion
    cephClusterID: 5c63ad7e-74fe-4724-a511-4ccdc560da56
```

Reply:
I think it should refer to this.

https://github.com/ceph/ceph-csi/blob/devel/cmd/cephcsi.go#L71 

https://github.com/rook/rook/blob/8dc88c39b105c79f8f5678f83dd23414411daa40/deploy/examples/operator.yaml#L59-L62

should this be overrideable at CephCSIDriver level ?

Reply:
This refers to the clusterName where cephcsi driver is running (typically openshift clusterID or some arbitrary kubernetes cluster name admin sets.

Reply:
https://github.com/rook/rook/blob/8dc88c39b105c79f8f5678f83dd23414411daa40/deploy/examples/operator.yaml#L59-L61 This is some configuration even Rook is not detecting it today and user is adding it. In openshift the clusterID is something specific to openshift CRD and in kubernetes we don't have any API for clusterID and its something user can choose alternative names they want for this field.

Reply:
What if this clusterName is not set? Since the setting today is not set by default for Rook upstream users I assume it's not required. It's only needed for some scenarios, right? Are those features disabled if not set?

Reply:
Yes you are correct, This is not a mandatory parameter, if its not set we will disable the feature in backend